pkg/controller/service/service_controller.go

```diff
@@ -241,6 +240,20 @@ func (s *ServiceController) processServiceUpdate(cachedService *cachedService, s
 			}
 		}
 	}
+
+	if cachedService.state != nil {
+		if !s.needsUpdate(cachedService.state, service) {
+			// The service does not require an update which means it was placed on the work queue
+			// by the node sync loop and indicates that the hosts need to be updated.
+			err := s.updateLoadBalancerHosts(service)
```

Comment:
Inferring source from state seems a little odd… does doing host updates when needsUpdate returns false mean this controller will do work on every service even when it should be in steady state?

Reply:
Services are only added to the queue under three conditions:
1. The service requires an update. That's determined by `needsUpdate` which is checked in the informer OnUpdate.
2. The node sync loop determined that hosts change from the last sync.
3. An error occurred so retry.

The `UpdateLoadBalancer` method is supposed to be cheap for cloud providers. It was added so that cloud providers could have a method that handled _only_ updating load balancer hosts. 

This approach should not (and doesn't from what I tell in my tests) add any extra calls from before.

Reply:
From https://github.com/kubernetes/kubernetes/issues/56443. I think the logic here is too hacky and may not cover all corner cases. What we really want here is to distinguish "service update" and "nodeSync update". The condition `!s.needsUpdate(cachedService.state, service)` is too broad that it includes also the retry case of "service update" (given how we cache service).

Besides, putting "nodeSync update" into the same work queue as "service update" might introduce another problem that one update could override the other. Ref https://github.com/kubernetes/kubernetes/issues/52495#issuecomment-330670308, within the working queue we don't save duplicate key, if both "nodeSync update" and  "service update" come in before anyone leaves the queue, it will end up with only one update (depends on how we decide what update it is). It seems to me that the working queue mechanism also needs to be adjusted before we can put in "nodeSync update".

cc @bowei

Reply:
> Besides, putting "nodeSync update" into the same work queue as "service update" might introduce another problem that one update could override the other

One update won't overwrite the other because the sync checks if the service needs an update. Both cloud provider calls (`EnsureLoadBalancer` and `UpdateLoadBalancer`) will update the hosts, so if both happen it will go with `EnsureLoadBalancer` which is what we want.

Reply:
As I mentioned in the other PR https://github.com/kubernetes/kubernetes/pull/56448#discussion_r153508837, the finalizer support will clean this all up and I think we should revert this until the finalizer PR cleans up all the cache/delete logic.

Reply:
> Both cloud provider calls (EnsureLoadBalancer and UpdateLoadBalancer) will update the hosts, so if both happen it will go with EnsureLoadBalancer

I think this assumption is inaccurate, ref the LoadBalancer interface, it doesn't explicitly define `EnsureLoadBalancer()` should update the hosts:
https://github.com/kubernetes/kubernetes/blob/e5aec8645d740123dfc8f1400b66bca05673b833/pkg/cloudprovider/cloud.go#L100-L109

And in fact, `EnsureLoadBalancer()` in GCE cloudprovider doesn't update hosts. Hence for GCE this is the case where "service update" overrides "nodeSync update".

Reply:
> creates a new load balancer 'name', or updates the existing one

Should include the backends (nodes). Why would it partially update the load balancer?

Reply:
@MrHohn from what I see GCE does make sure that backends are up-to-date on EnsureLoadBalancer calls via [this method](https://github.com/kubernetes/kubernetes/blob/e57accbd03b1265340fdd266d0afe2b7a8e1c171/pkg/cloudprovider/providers/gce/gce_loadbalancer_internal.go#L408), which is called by [`ensureInternalLoadBalancer`](https://github.com/kubernetes/kubernetes/blob/e57accbd03b1265340fdd266d0afe2b7a8e1c171/pkg/cloudprovider/providers/gce/gce_loadbalancer_internal.go#L37) and `ensureExternalLoadBalancer` does as well via chekcing first [this method](https://github.com/kubernetes/kubernetes/blob/e57accbd03b1265340fdd266d0afe2b7a8e1c171/pkg/cloudprovider/providers/gce/gce_loadbalancer_external.go#L675).

Reply:
Thanks for gathering the links, I didn't look into the internal one before, seems like it does check for hosts update. Though ATM the external one doesn't check for hosts update.
https://github.com/kubernetes/kubernetes/blob/e57accbd03b1265340fdd266d0afe2b7a8e1c171/pkg/cloudprovider/providers/gce/gce_loadbalancer_external.go#L675-L676

Reply:
Ah, missed that comment. The external one is definitely more complex than the internal one 👼 

It sounds like we should definitely revert this then and we need to come to an agreement on what `EnsureLoadBalancer` and `UpdateLoadBalancer` should do for each cloud provider. It was my understanding that `EnsureLoadBalancer` should completely update the load balancer, which is how AWS, Azure, Oracle & DigitalOcean handle it (only ones i checked). cc @wlan0 @luxas 

I'll open a PR to revert this for 1.9 @MrHohn